compiler/pipes/final-check.cpp

```diff
@@ -107,6 +109,97 @@ void check_to_array_debug_call(VertexAdaptor<op_func_call> call) {
   }
 }
 
+void check_json_fields_duplication(ClassPtr klass) noexcept {
+  std::set<std::string_view> used_json_keys;
+
+  for (auto ancestor : klass->get_all_ancestors()) {
+    ancestor->members.for_each([&used_json_keys, ancestor](const ClassMemberInstanceField &field) {
+      auto rename = field.kphp_json_tags ? field.kphp_json_tags->get_rename_value() : vk::string_view{};
+      auto key = rename.empty() ? field.local_name() : rename;
+      bool success = used_json_keys.emplace(key).second;
+      kphp_error_return(success, fmt_format("@kphp-json field '{}' met twice, first time in class '{}'", key, ancestor->name));
+    });
+  }
+}
+
+void check_json_tag_correctness(ClassPtr klass) noexcept {
+  klass->members.for_each([](const ClassMemberInstanceField &field) {
+    const auto type = tinf::get_type(field.var)->ptype();
+    if (field.kphp_json_tags && field.kphp_json_tags->has_array_as_hashmap()) {
+      kphp_error_return(type == tp_array,
+        fmt_format("@kphp-json array_as_hashmap tag is allowed only above array type, got above {} field", field.local_name()));
+    }
+    if (field.kphp_json_tags && field.kphp_json_tags->is_raw_string()) {
+      kphp_error_return(type == tp_string,
+        fmt_format("@kphp-json raw_string tag is allowed only above string type, got above {} field", field.local_name()));
```

Comment:
```suggestion
        fmt_format("@kphp-json raw_string tag is allowed only above string type, got above ${} field", field.local_name()));
```

Reply:
Hm, I better leave as is, as all comments below.